### PR TITLE
[#203] New microservice that implements the device registry - so far …

### PR DESCRIPTION
**Pull request by sysexcontrol**
…the credentials API is implemented as first step.

As a proposal I made a new Maven module "services" under which the maven module "device-registry" is hosted. 
Alternatively the device-registry could be at the top level as well - to me it seems to be a cleaner structure with the "services" module. Feedback welcome of course - if agreed on, the auth-server might move there as well (?).

Steps to follow include:
- integration tests
- docker swarm setup
- registration API
- REST Adapter using the device registry instead of Hono for the registration API

Signed-off-by: Karsten Frank <Karsten.Frank@bosch-si.com>

### CODE REVIEW

File: services/device-registry/pom.xml [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+                    <assembly>
+                      <mode>dir</mode>
+                      <basedir>/</basedir>
+                      <inline>
+                        <fileSets>
+                          <fileSet>
+                            <directory>${project.basedir}/target/classes/certs</directory>
+                            <outputDirectory>etc/hono/certs</outputDirectory>
+                            <includes>
+                              <include>devreg-server-*.pem</include>
```
Comment (calohmn) [inline services/device-registry/pom.xml]: I think this should be `device-registry-*.pem`.

**Comment by sophokles73** [comment]
> As a proposal I made a new Maven module "services" under which the maven module "device-registry" is hosted.
Alternatively the device-registry could be at the top level as well - to me it seems to be a cleaner structure with the "services" module. Feedback welcome of course - if agreed on, the auth-server might move there as well (?).

That makes a lot of sense to me :-)

File: services/device-registry/pom.xml (hunk 1 above)
Comment (sysexcontrol) [inline services/device-registry/pom.xml]: Thx! Is fixed in latest push.

**Comment by sophokles73** [comment]
@sysexcontrol ,

sorry to say this but I have re-created all certificates as part of #209. Can you rebase your PR and check if you can use the existing certs and keys?

**Comment by sysexcontrol** [comment]
Rebased to current master, pushed again.
Please have a look, device registration API will follow next then.

File: services/device-registry/pom.xml [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+                    </ports>
+                    <runCmds>
+                      <runCmd>adduser -S -D hono nogroup</runCmd>
+                    </runCmds>
+                    <user>hono</user>
+                    <workdir>/home/hono</workdir>
+                    <cmd>
+                      <exec>
+                        <arg>java</arg>
+                        <arg>-Dvertx.disableDnsResolver=true</arg>
```
Comment (sophokles73) [inline services/device-registry/pom.xml]: you can safely remove these three lines regarding DNS. We are now using the vert.x hostname resolver (which needs to be configured explicitly, though). See `ApplicationConfing` from the hono-app.

File: services/device-registry/readme.md [hunk 3]
```diff
@@ -0,0 +1,8 @@
+This module contains an implementation of Hono's Device Registry as a separated microservice.
```
Comment (sophokles73) [inline services/device-registry/readme.md]: `.. as a separate ...`

File: services/device-registry/src/main/java/org/eclipse/hono/service/deviceregistry/impl/Application.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Bosch Software Innovations GmbH - initial creation
+ */
+
+package org.eclipse.hono.service.deviceregistry.impl;
```
Comment (sophokles73) [inline services/device-registry/src/main/java/org/eclipse/hono/service/deviceregistry/impl/Application.java]: I would prefer `org.eclipse.hono.deviceregistry` as a top level package.

File: services/device-registry/src/main/java/org/eclipse/hono/service/deviceregistry/impl/ApplicationConfig.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+    private static final Vertx vertx = Vertx.vertx();
+
+    /**
+     * Gets the singleton Vert.x instance to be used by Hono.
+     * 
+     * @return the instance.
+     */
+    @Bean
+    public Vertx vertx() {
```
Comment (sophokles73) [inline services/device-registry/src/main/java/org/eclipse/hono/service/deviceregistry/impl/ApplicationConfig.java]: see `ApplicationConfig` in hono-app for how to configure vert.x DNS properly.

File: services/device-registry/src/main/java/org/eclipse/hono/service/deviceregistry/impl/ApplicationConfig.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        return factory;
+    }
+
+    /**
+     * Exposes this service's configuration properties as a Spring bean.
+     * 
+     * @return The properties.
+     */
+    @Bean
+    @ConfigurationProperties(prefix = "hono.deviceRegistry")
```
Comment (sophokles73) [inline services/device-registry/src/main/java/org/eclipse/hono/service/deviceregistry/impl/ApplicationConfig.java]: IMHO the prefix should not contain camel case segments, i.e. use `hono.device.registry` instead.

File: services/device-registry/src/main/java/org/eclipse/hono/service/deviceregistry/impl/SimpleDeviceRegistryServer.java [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+        connection.disconnect();
+    }
+
+    /**
+     * Gets the authenticated client principal name for an AMQP connection.
+     *
+     * @param con the connection to read the user from
+     * @return the user associated with the connection or {@link Constants#SUBJECT_ANONYMOUS} if it cannot be determined.
+     */
+    private String getUserFromConnection(final ProtonConnection con) {
```
Comment (sophokles73) [inline services/device-registry/src/main/java/org/eclipse/hono/service/deviceregistry/impl/SimpleDeviceRegistryServer.java]: replace with `org.eclipse.hono.utils.Constants.getClientPrincipal(ProtonConnection)`

**Comment by sysexcontrol** [comment]
No, your are absolutely right - in this PR there are no endpoints yet.
Both endpoints contained in the next PR - this PR makes the credentials standalone test running in the new microservice only.

File: services/device-registry/pom.xml (hunk 2 above)
Comment (sysexcontrol) [inline services/device-registry/pom.xml]: Thanks, will substitute it in ApplicationConfig.

File: services/device-registry/src/main/java/org/eclipse/hono/service/deviceregistry/impl/Application.java (hunk 4 above)
Comment (sysexcontrol) [inline services/device-registry/src/main/java/org/eclipse/hono/service/deviceregistry/impl/Application.java]: ok, will move it to top level.

File: services/device-registry/src/main/java/org/eclipse/hono/service/deviceregistry/impl/SimpleDeviceRegistryServer.java (hunk 7 above)
Comment (sysexcontrol) [inline services/device-registry/src/main/java/org/eclipse/hono/service/deviceregistry/impl/SimpleDeviceRegistryServer.java]: makes sense. Will replace it.

File: services/device-registry/src/main/java/org/eclipse/hono/service/deviceregistry/impl/ApplicationConfig.java (hunk 6 above)
Comment (sysexcontrol) [inline services/device-registry/src/main/java/org/eclipse/hono/service/deviceregistry/impl/ApplicationConfig.java]: yes, will change it.

File: services/device-registry/src/main/java/org/eclipse/hono/service/deviceregistry/impl/ApplicationConfig.java (hunk 5 above)
Comment (sysexcontrol) [inline services/device-registry/src/main/java/org/eclipse/hono/service/deviceregistry/impl/ApplicationConfig.java]: +1

File: services/device-registry/pom.xml [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+                <goals>
+                  <goal>build</goal>
+                </goals>
+              </execution>
+            </executions>
+          </plugin>
+          <plugin>
+            <groupId>io.fabric8</groupId>
+            <artifactId>fabric8-maven-plugin</artifactId>
+            <version>3.2.14</version>
```
Comment (sophokles73) [inline services/device-registry/pom.xml]: please use plugin configuration from hono-parent POM

File: services/device-registry/src/main/java/org/eclipse/hono/deviceregistry/impl/DeviceRegistryConfigProperties.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+import org.eclipse.hono.config.ServiceConfigProperties;
+import org.springframework.core.io.ClassPathResource;
+import org.springframework.core.io.Resource;
+
+
+/**
+ * Configuration properties for the device registry as own server.
+ *
+ */
+public class DeviceRegistryConfigProperties extends ServiceConfigProperties {
```
Comment (sophokles73) [inline services/device-registry/src/main/java/org/eclipse/hono/deviceregistry/impl/DeviceRegistryConfigProperties.java]: this should probably be final, shouldn't it. At least this class is not supposed to be extended by client code AFAIC

File: services/device-registry/src/main/java/org/eclipse/hono/deviceregistry/impl/FileBasedCredentialsService.java [hunk 10]
```diff
@@ -44,108 +51,82 @@
 @Repository
 @ConfigurationProperties(prefix = "hono.credentials")
 @Profile({"default", "credentials-file"})
-public class FileBasedCredentialsService extends BaseCredentialsService {
+public class FileBasedCredentialsService extends BaseCredentialsService<DeviceRegistryConfigProperties> {
```
Comment (sophokles73) [inline services/device-registry/src/main/java/org/eclipse/hono/deviceregistry/impl/FileBasedCredentialsService.java]: final ?

**Comment by sysexcontrol** [comment]
All changes reflected in recent push, can you have a look again, please?
FYI: since there was a merge conflict, I rebased the branch as well.

File: services/device-registry/src/main/java/org/eclipse/hono/deviceregistry/impl/FileBasedCredentialsService.java [hunk 11]
```diff
@@ -44,108 +51,82 @@
 @Repository
 @ConfigurationProperties(prefix = "hono.credentials")
```
Comment (sophokles73) [inline services/device-registry/src/main/java/org/eclipse/hono/deviceregistry/impl/FileBasedCredentialsService.java]: there are no additional properties to be set, are there? You are using the `DeviceRegistryConfigProperties` as indicated by the class parameter ...

File: services/device-registry/src/main/java/org/eclipse/hono/deviceregistry/impl/SimpleDeviceRegistryServer.java [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+
+/**
+ * A server that implements Hono's Credentials and Registration APIs as the default implementation of Hono's device registry.
+ *
+ * TODO: Currently only implements the Credentials API. Registration API implementation to be added.
+ * TODO: add metrics.
+ */
+@Component
+@Scope("prototype")
+public class SimpleDeviceRegistryServer extends AmqpServiceBase<DeviceRegistryConfigProperties> {
```
Comment (sophokles73) [inline services/device-registry/src/main/java/org/eclipse/hono/deviceregistry/impl/SimpleDeviceRegistryServer.java]: final ?

File: services/device-registry/src/main/java/org/eclipse/hono/deviceregistry/impl/DeviceRegistryConfigProperties.java (hunk 9 above)
Comment (sysexcontrol) [inline services/device-registry/src/main/java/org/eclipse/hono/deviceregistry/impl/DeviceRegistryConfigProperties.java]: +1

**Comment by sysexcontrol** [comment]
Reflected changes, rebased again (to enable the use of current hono-parent). Please look again.

File: services/device-registry/src/test/java/org/eclipse/hono/service/credentials/StandaloneCredentialsApiTest.java [hunk 13]
```diff
@@ -9,48 +9,52 @@
  * Contributors:
  *    Bosch Software Innovations GmbH - initial creation
  */
-package org.eclipse.hono.server;
+package org.eclipse.hono.service.credentials;
```
Comment (sophokles73) [inline services/device-registry/src/test/java/org/eclipse/hono/service/credentials/StandaloneCredentialsApiTest.java]: package `org.eclipse.hono.deviceregistry`

File: services/device-registry/src/main/java/org/eclipse/hono/deviceregistry/impl/FileBasedCredentialsService.java [hunk 14]
```diff
@@ -10,7 +10,9 @@
  * Bosch Software Innovations GmbH - initial creation
  */
 
-package org.eclipse.hono.service.credentials.impl;
+package org.eclipse.hono.deviceregistry.impl;
```
Comment (sysexcontrol) [inline services/device-registry/src/main/java/org/eclipse/hono/deviceregistry/impl/FileBasedCredentialsService.java]: I moved it to org.eclipse.hono.deviceregistry as well. 

As soon as the registration API is contained (next PR) the other option is better IMHO:
- service.credentials.impl for FileBasedCredentialsService.java
- service.registration.impl for FileBasedRegistrationService.java
WDYT?

**Comment by sysexcontrol** [comment]
Please check again, thx.

File: services/device-registry/src/main/java/org/eclipse/hono/deviceregistry/ApplicationConfig.java [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+import io.vertx.core.Vertx;
+
+/**
+ * Spring Boot configuration for the simple device registry server application.
+ *
+ */
+@Configuration
+public class ApplicationConfig {
+
+    private static final Vertx vertx = Vertx.vertx();
```
Comment (sophokles73) [inline services/device-registry/src/main/java/org/eclipse/hono/deviceregistry/ApplicationConfig.java]: obsolete

File: services/device-registry/src/test/java/org/eclipse/hono/service/credentials/StandaloneCredentialsApiTest.java (hunk 13 above)
Comment (sophokles73) [inline services/device-registry/src/test/java/org/eclipse/hono/service/credentials/StandaloneCredentialsApiTest.java]: still the wrong package, isn't it?

**Comment by sysexcontrol** [comment]
Pushed changes, please look again, thx!

**Comment by sophokles73** [comment]
@sysexcontrol can you please rebase and squash?

**Comment by sysexcontrol** [comment]
Rebased and squashed.

**Pull request merged**